src/cache-utils.ts

```diff
@@ -70,9 +78,16 @@ export const getCacheDirectoryPath = async (
   packageManagerInfo: PackageManagerInfo,
   packageManager: string
 ) => {
-  const stdOut = await getCommandOutput(
-    packageManagerInfo.getCacheFolderCommand
-  );
+  let stdOut = await getCommandOutput(packageManagerInfo.getCacheFolderCommand);
+
+  // pnpm returns 'undefined' if no custom store path is set
+  if (stdOut === 'undefined') {
+    stdOut = '';
+  }
+
+  if (!stdOut && packageManagerInfo.defaultCacheFolder) {
+    stdOut = packageManagerInfo.defaultCacheFolder;
```

Comment:
I think we should extend this function a bit and add validation check that retrieved cache path exists on disk.

The `toolkit/cache` package doesn't handle this situation properly and if cache path doesn't exist on disk, it won't fail. It will create malformed archive with size 22 B silently and then fail to restore such cache. We had similar issue with Yarn recently: https://github.com/actions/setup-node/issues/281 .

Since we fallback to `~/.pnpm-store` in case of issue, it would be great to add such validation to check if this folder exists and fail earlier if it doesn't exist.
I suggest something like this:
```typescript
if (!stdOut && packageManagerInfo.defaultCacheFolder) {
  stdOut = packageManagerInfo.defaultCacheFolder;
}

if (!stdOut) {
  throw new Error(`Could not get cache folder path for ${packageManager}`);
}
core.debug(`${packageManager} path is ${stdOut}`);

if (!fs.existsSync(stdOut)) {
  throw new Error(`Cache folder path is retrieved for ${packageManager} but doesn't exist on disk: ${stdOut}`);
}

return stdOut;
```

The single problem that `fs.exists` won't handle `~` in path properly but we can replace `defaultCacheFolder: '~/.pnpm-store'` with `defaultCacheFolder: path.join(os.homedir(), '.pnpm-store')`
Please let me know if you see any alternative options

Reply:
This makes sense to me, I have updated with your suggestion 👍

Reply:
Thanks, but some tests fail. I think we should also mock `fs.existsSync` to return `true` in tests.

Reply:
I think the test failures are legitimate, this function is used to compute the cache path when _restoring_ the cache as well. Do you have any preferences on how to proceed? I think we only want to throw an error on the save path, not on the restore path.

Reply:
I moved the `if (!fs.existsSync(cachePath))` check to `cache-save.ts`, that way it only throws that error when attempting to save the cache.

Reply:
Personally, I would prefer to keep this check in `getCacheDirectoryPath` function. When we calculate cache path from tool - we check that this path is correct.
Looks like some tests still fail. Have you tried to run it locally? Does it work for you? I guess we still need to mock `fs.existSync` invocation.

Reply:
@maxim-lobanov I don't believe the error can stay in `getCacheDirectoryPath` because of this:

https://github.com/actions/setup-node/blob/38d90ce44d5275ad62cc48384b3d8a58c500bb5f/src/cache-restore.ts#L21-L33

The directory name returned by `getCacheDirectoryPath` is used to compute the cache path _before_ it's restored, so `fs.existsSync` would always fail, as the cache is not restored at that point. That's why I think the check only applies to `cache-save.ts`, as it's only the save path that needs to check the existence of the directory.

I'll work on mocking `fs.existsSync` for the cache save tests...

Reply:
Okay, `fs.existsSync` is mocked, all tests are passing for me locally ✅

Reply:
Hm, auth tests fail in CI for some reason. May be mocks are not clean up properly? We can help with investigation.

Also, I would like to discuss one more issue: 
`defaultCacheFolder: path.join(os.homedir(), '.pnpm-store')` doesn't work for Windows (e2e CI fails on this platform).
I have tried to experiment and looks like pnpm uses different cache dir for different disk drives. The following commands:
```
New-Item -ItemType Directory -Path "C:\Temp2"
cd "C:\Temp2"
pnpm store path
          
New-Item -ItemType Directory -Path "D:\Temp2"
cd "D:\Temp2"
pnpm store path
```
produces the following output:
```
C:\Users\runneradmin\.pnpm-store\v3
D:\.pnpm-store\v3
```

Looks like cache path on Windows depends on current directory so we can't simply set `defaultCacheFolder`.

I wonder if we can just rely on `pnpm store path` command introduced by you without fallback path and note in README that caching for pnpm works for pnpm 6.10+
We don't have much experience with pnpm so we would like to hear your opinion.

Reply:
I'm comfortable with relying on `pnpm store path`, now that it's released. I'll push that change now, so that we can see if tests will pass. (I'll also set the `fs` mock to unmock after the cache save tests)

An alternative or addition would be to import the [`@pnpm/store-path` package](https://www.npmjs.com/package/@pnpm/store-path), which allows you to programatically resolve the store path. I think it would be best used as a fallback for if `pnpm store path` failed. However, I assume that you'd prefer to not include another dependency in this action, and would prefer to avoid package-manager-specific integrations. But let me know if I'm incorrect about that.

Reply:
I like the current approach and I think we should proceed with it for now. We can revisit in future if users really want to use older versions of pnpm but I wouldn't add new dependencies here.

Looks like all tests are green, thank you!